docs/guides/napari_models.md

```diff
@@ -0,0 +1,229 @@
+    * the code for this is private and lives in `napari/_vispy`
+
+The separation of the python models from viewer GUI code allows:
+
+* analysis plugins to be developed without worrying about the GUI
+  aspect
+* napari to have the option to move away from the rendering backend currently
+  used
+* tests to be easily run headlessly
+* the python models to be run headlessly (see
+  [Running napari headlessly](../howtos/headless.md) for more)
+
+## Python models and events
+
+Commonly, python models in napari are classes that store information about their
+state as an attribute and are the "source of ground truth". When these
+attributes are changed an "event" needs to be emitted such that relevant
+observers of the model (such as other classes) can take the appropriate
+action.
+
+One way this is achieved in napari is via getters and setters. Let's take
+for example the `Dims` class with a selected few attributes:
+
+```python
+from napari.utils.events import EmitterGroup
+
+class Dims:
+    """Dimensions object modeling slicing and displaying.
+
+    Parameters
+    ----------
+    ndim : int
+        Number of dimensions.
+    ndisplay : int
+        Number of displayed dimensions.
+    ...
+    """
+    def __init__(self, ndim, ndisplay):
+        self._ndim_ = ndim
+        self._ndisplay = ndisplay
+
+        # an `EmitterGroup` manages a set of `EventEmitters`
+        # we add one emitter for each attribute we'd like to track
+        self.events = EmitterGroup(source=self, ndim=None, ndisplay=None)
+
+    # for each attribute, we create a `@property` getter/setter
+    # so that we can emit the appropriate event when that attribute
+    # is changed using the syntax: ``Dim.attribute = new_value``
+    @property
+    def ndim(self):
+        """Number of dimensions."""
+        return self._ndim
+
+    @ndim.setter
+    def ndim(self, value):
+        self._ndim = value
+        # emit the ndim "changed" event
+        self.events.ndim(value=value)
+
+    @property
+    def ndisplay(self):
+        """Number of displayed dimensions."""
+        return self._ndisplay
+
+    @ndisplay.setter
+    def ndisplay(self, value):
+        self._ndisplay = value
+        # emit the ndisplay "changed" event
+        self.events.ndisplay(value=value)
+```
+
+Another object can then "listen" for changes in our `Dims` model and register
+a callback function with the event emitter of the attribute they would like
+to watch:
+
+```python
+# create an instance of the model
+dims = Dims(ndim=3, ndisplay=2)
+
+# define some callback that should respond to changes in the model
+def _update_display(self):
+    """
+    Updates display for all sliders.
+    """
+    # the code updating the display code is not relevant for this
+    # example thus has been ommited.
+    nsteps = self.dims.nsteps
+    print(f"Update number of dimensions displayed to {nsteps}")
+
+# register our callback with the model
+dims.events.ndisplay.connect(_update_display)
```

Comment:
I'm not sure which level you are targeting that for, but there is a number of implicit things in the black of code. There is a mix of standalone function and objects method. 
As `_update_display(...)` take self, then we can assume that this will be an object method.  Is mention  `self.dims.nsteps`, so I assume that `dims = Dims(...)` has m been assigned to the `self.dims` attribute, and thus I believe that `connect(_update_display)` should connect  to `self._update_display` somehow, otherwise the first argument that _update_displauy will get will be an `Event` with attributes like `source`, `value` ...

`nsteps` also don't exists on `Dims`.



```suggestion
dims = Dims(ndim=3, ndisplay=2)

# define some callback that should respond to changes in the model
# if the function takes a single parameter it will receive the event
# as first value. 
def _update_display(event):
    """
    Updates display for all sliders.
  
    """
    # the code updating the display code is not relevant for this
    # example thus has been omitted.
    
    # we can get the source object of the event, 
    assert event.source == dims 
    
    # ... and query any attributes
    ndisplay = event.source.ndisplay
    
    # ... or directly get the new value for this specific event:
    assert ndisplay == event.value
    print(f"Update number of dimensions displayed to {ndisplay}")

# register our callback with the model
dims.events.ndisplay.connect(_update_display)
```

Reply:
> There is a mix of standalone function and objects method.

Thanks, mostly I was trying to give an easy example from the napari code base but I realise now that it was a method and did not make sense how I had it.

Reply:
Thanks, the new version reads much better IMHO, thanks a lot for doing all that.